clients/admin-ui/src/features/common/CheckboxTree.tsx

```diff
@@ -1,5 +1,56 @@
-import { Box, Checkbox } from "@fidesui/react";
-import { Fragment, ReactNode } from "react";
+/**
+ * This CheckboxTree does a few opinionated things to facilitate DataCategory selection
+ *   * Indeterminate parents when not all children are selected
+ *   * Separate "selected" and "checked" states, since only the most specific child is "selected".
+ *     * However, the parents are not "selected"—they will either be "indeterminate", or "checked" depending
+ *       on if the siblings are also "selected"
+ *   * "Selected" children render expanded up until the child
+ */
+
+import { Box, Checkbox, IconButton } from "@fidesui/react";
+import { Fragment, ReactNode, useEffect, useState } from "react";
+
+import { ArrowDownLineIcon, ArrowUpLineIcon } from "./Icon";
+
+export const getAncestorsAndCurrent = (nodeName: string) => {
+  const splitNames = nodeName.split(".");
+  const ancestors: string[] = [];
+  splitNames.forEach((name) => {
+    const parent =
+      ancestors.length > 0 ? ancestors[ancestors.length - 1] : null;
+    if (!parent) {
+      ancestors.push(name);
```

Comment:
these functions are pure JS so they don't know about react state (they're just helper functions, though I included them right in the component since they are tied closely to the CheckboxTree). I could probably do something like:

```
let ancestors = [];
...
ancestors = [...ancestors, name]
```

but I feel like `.push` is easier to follow and also doesn't need to use a `let`/create new array objects each time. But let me know if you feel differently, happy to learn!

Reply:
Since we're overall in a React ecosystem though and these helpers exist in that context, I think it makes sense to follow React best-practices when altering lists/arrays, even in helper functions. In general, you don't want to alter the state directly in React. [Thou Shalt Not Alter State Directly.

That's pretty much a commandment-level dictum in React. Every time you want to change it you should use setState() handing it the new info to put there. Since push() adds to the existing array instead of returning a new one it goes against that principle.

[React doc](https://facebook.github.io/react/docs/state-and-lifecycle.html#using-state-correctly)
[StackOverflow discussion](https://stackoverflow.com/questions/26253351/correct-modification-of-state-arrays-in-react-js)

I think doing what you suggested above would make it easier to follow in a React context and also make it less error-proned. I don't mind pushback here if you disagree though - also happy to learn if you think the above doesn't apply.

Reply:
I totally agree not to do it on React state, which is what both of those links are talking about. This case is more of "iterate through a list and add certain items to an array", so push makes more sense to me, especially since we don't have access to any of the React state to work with. I do agree in general pushes/pops generally aren't needed since you can mostly get by with `filter` and `map` but this was a case where I thought it was more readable to go `push` since the recursive logic is already confusing

Reply:
If this isn't ultimately dealing with any React state directly, then I'm comfortable approving. Anyway, I think I'm allergic to push stylistically (and because I have direct-mutation worries), so I do prefer the spread syntax you listed above 😆 But if push is easier to follow, I'll stop grouching about it.

Reply:
heh fair enough!